### DISCUSSION TITLE
Welcome to SMA-Handshake-Eng-Software-2022.1- Discussions!

### DISCUSSION THREAD

**Discussion by gaprilus** [ES-UFABC/SMA-Handshake-Eng-Software-2022.1-#20] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by gaprilus** [comment]
import React, {Component} from 'react'
import servicesCadastro from '../middleware/servicesCadastro'

export default class Cadastro extends Component {


    state = {
        cadastro:{
            tipo:'',
            nome:'',
            cpf_cnpj:'',
            telefone: '',
            estado: '',
            cidade: '',
            cep: '',
            logradouro:'' ,
            numero: '',
            email: '',
            senha: '',
        },
        checkSenha: '',
    }

    handleBlur = (e) => {
        const {cadastro} = this.state
        let newCadastro = cadastro

        newCadastro[e.target.id] = e.target.value
        this.setState({cadastro: newCadastro});
        console.log(cadastro)

    }
    handleSubmit = () => {
        
        const {cadastro} = this.state

        const novoCadastro = new servicesCadastro ()
        novoCadastro.setCadastro(cadastro)

    }

    render() {
        return (

            <>
            <div>
            </div>
            <div className='div-cadastro'>
                <form className='cadastro' onSubmit={this.handleSubmit}>
                <h1>Cadastro</h1>
                  <div className='div-global'>
                  <div className='div-form'>
                    <label>Tipo</label>
                    <input list="tipos" id='tipo' onBlur={this.handleBlur}/>
                    <datalist id="tipos">
                    <option value="Produtor" />
                    <option value="Consumidor" />
                    </datalist>
                    <label>Nome</label>
                    <input id='nome' onBlur={this.handleBlur}></input>
                    <label>CPF/CNPJ</label>
                    <input id = 'cpf_cnpj' onBlur={this.handleBlur}></input>
                    <label>Telefone</label>
                    <input id='telefone' onBlur={this.handleBlur}></input>
                    <label>Estado</label>
                    <input id='estado' onBlur={this.handleBlur}></input>
                    <label>Cidade</label>
                    <input id='cidade' onBlur={this.handleBlur}></input>
                    </div>
                    <div className='div-form'>
                    <label>CEP</label>
                    <input id='cep' onBlur={this.handleBlur}></input>
                    <label>Logradouro</label>
                    <input id='logradouro' onBlur={this.handleBlur}></input>
                    <label>Nº</label>
                    <input id='numero' onBlur={this.handleBlur}></input>
                    <label>E-mail</label>
                    <input id='email' type='email' onBlur={this.handleBlur}></input>
                    <label>Senha</label>
                    <input id='senha' type='password' onBlur={this.handleBlur}></input>
                    <label>Confirmar Senha</label>
                    <input id='checkSenha' type='password' onBlur={this.handleBlur}></input>
                    </div>
                   <button className='confirmar-cadastro'>Cadastrar-se</button>
                    </div>
                </form>
            </div>
            </>

        )
    }

}

**Comment by gaprilus** [comment]
import React, {Component} from 'react'

export default class Navbar extends Component {


    state = {
        login:{
        email: '',
        senha: ' ',
        }
    }

    handleBlur = (e) => {
        const {login} = this.state
        let newLogin = login

        newLogin[e.target.id] = e.target.value
        this.setState({login: newLogin});
        console.log(login)
    }
    handleSubmit = () => {
        
    }

    render() {
        return (

            <>
            <header>
            <div>
                <h1><a href='/'>SMA Handshake</a></h1>
            </div>
            <div className='login-div'>
                <form onSubmit={this.handleSubmit}>
                <label>Usuário</label>
                <input id='email' onBlur={this.handleBlur}></input>
                <label>Senha</label>
                <input id='senha' onBlur={this.handleBlur}></input>
                <button >Entrar</button>
                </form>
                <a href='cadastro'>Cadastrar-se</a>
            </div>
            </header>
            </>

        )
    }

}

**Comment by gaprilus** [comment]
import React, {Component} from 'react';

import DataGrid, {
  Column,
  Editing,
  Popup,
  Paging,
  Lookup,
  Form,
  Item,
  Button,
} from 'devextreme-react/data-grid';
import {Form as FormStd, SimpleItem, ButtonItem, GroupItem} from 'devextreme-react/form';
import { Popup as StdPopup, Position, ToolbarItem,   } from 'devextreme-react/popup';


const data = [{
    id: 1,
    commodity: 'soja',
    disponibilidade: '2022/05/01',
    quantidade: 10,
    preco: 90,

}]

const commoditiesOptions = { items: ['Milho','Soja','Trigo'], searchEnabled: true, value: '' };
const statesOptions = { items: ['São Paulo','Bahia','Rio de Janeiro'], searchEnabled: true, value: '' };
const popupButtonOptions ={ text: 'Comprar', width: 100}

const buttonOptions={
    text: 'Buscar',
    type: 'success',
    useSubmitBehavior: true,
    width: 200
  }

export default class Consumidor extends Component {
    
    state = {
        busca_dados: {
        commodity: '',
        estado: '',
    },
    visiblePopup: false,
    }
    


    handleSubmit = (e) => {
        e.preventDefault()
        console.log(e)
        console.log(this.state.busca_dados)
    }

    handleChange = (e) => {

        const {busca_dados} = this.state
        let newDados = busca_dados

        console.log(e.value)
       
        newDados[e.dataField] = e.value
        this.setState({busca_dados: newDados});
    
    }

    handlePurchase = (e) => {
        console.log('Compra!')

        this.setState({visiblePopup:true})

        console.log(this.state.visiblePopup)
    }

    handleHide = (e) => {
        console.log('Close!')

        this.setState({visiblePopup:false})

        console.log(this.state.visiblePopup)
    }

    render() {

        const busca_dados = this.state.busca_dados;
    
        return (

            <>
            <div>
                <div className='std-div'>
            <h1>Perfil do Consumidor</h1>
                    <form onSubmit={this.handleSubmit}>
                <FormStd
                  width={800}
                    colCount={1}
                    formData={busca_dados}
                    onFieldDataChanged={this.handleChange}>
                        <GroupItem caption='Busca' colCount={3} >
                      
                        <SimpleItem dataField='commodity'
                            editorType='dxSelectBox' 
                            editorOptions={commoditiesOptions}/>
                        <SimpleItem dataField='estado' editorType='dxSelectBox' editorOptions={statesOptions}></SimpleItem>
                        <ButtonItem 
                        buttonOptions={buttonOptions}
                        >

                        </ButtonItem>
                       
                    </GroupItem>
                </FormStd>
                </form>
                </div>
            </div>
            <div className='std-div'>
            
                <DataGrid
                    dataSource={data}
                   >
                       
                    <Paging enabled={false} />
                        <Column dataField='commodity' />
                        <Column dataField='disponibilidade' />
                        <Column dataField='quantidade' />
                        <Column dataField='preco' />
                        <Column type='buttons'>
                            <Button 
                                text='Comprar' 
                                icon='money'
                                onClick={this.handlePurchase}
                                 />
                        </Column>
                      
                            <Form>
                            
                                <Item dataField= 'commodity'/>
                                <Item dataField ='disponibilidade' />
                                <Item dataField = 'quantidade' />
                                <Item dataField = 'preco' />
                            </Form> 
                </DataGrid>
            </div>

            <StdPopup
                visible={this.state.visiblePopup}
                onHiding={this.handleHide}
                showCloseButton={true}
                // closeOnOutsideClick={this.handleHide}
                showTitle={true}
                title='Compra'
                width={420}
                height={250}
                >
                    
                    <ToolbarItem text='Quantidade'
                    toolbar="bottom"
                    location="center"/>
                    <ToolbarItem widget='dxTextBox'
                    toolbar="bottom"
                    location="center"/>
                    <ToolbarItem widget='dxButton'
                    options={popupButtonOptions}
                    toolbar="bottom"
                    location="center"/>

                    <p>Preço: <span>{data[0].preco}</span></p>
                    <p>Valor: <span>{data[0].preco}</span></p>
                    
                
            </StdPopup>
            </>

        

        )
    }

}

**Comment by gaprilus** [comment]
import React, {Component} from 'react';

import DataGrid, {
  Column,
  Editing,
  Popup,
  Paging,
  Lookup,
  Form,
  Item,
  Button,
  Selection,
} from 'devextreme-react/data-grid';
import {Form as FormStd, SimpleItem, ButtonItem, GroupItem} from 'devextreme-react/form';
import { Popup as StdPopup, Position, ToolbarItem, Item as PopupItem   } from 'devextreme-react/popup';

import {setOrdem} from '../middleware/servicesConsumidor'


const columns = ['id', 'commodity','disponibilidade', 'quantidade','preco','comprar']

const data = [{
    id: 1,
    commodity: 'soja',
    disponibilidade: '2022/05/01',
    quantidade: 10,
    preco: 90,
    comprar: 0,
    // executar: false,
    

}]

const commoditiesOptions = { items: ['Milho','Soja','Trigo'], searchEnabled: true, value: '' };
const statesOptions = { items: ['São Paulo','Bahia','Rio de Janeiro'], searchEnabled: true, value: '' };

const popupButtonOptions ={ 
    text: 'Comprar', 
    width: 100,
}

const buttonOptions={
    text: 'Buscar',
    type: 'success',
    useSubmitBehavior: true,
    width: 200
  }



export default class Consumidor extends Component {
    
  
    state = {
        busca_dados: {
        commodity: '',
        estado: '',
    },
    visiblePopup: false,
    quantidade: 0,
    }

    popupQtyOptions={
        
        onDataFieldChange: this.handleQtyChange,
        
      }
    
handleQtyChange = (e) => {
    console.log(e.value)
}

    handleSubmit = (e) => {
        e.preventDefault()
        console.log(e)
        console.log(this.state.busca_dados)
    }

    handleChange = (e) => {

        const {busca_dados} = this.state
        let newDados = busca_dados

        console.log(e.value)
       
        newDados[e.dataField] = e.value
        this.setState({busca_dados: newDados});
    
    }

    handlePurchase = (e) => {
        console.log('Compra!')

        this.setState({visiblePopup:true})
    
        
        console.log(this.state.visiblePopup)
    }

    handleHide = (e) => {
        console.log('Close!')

        this.setState({visiblePopup:false})

        console.log(this.state.visiblePopup)
    }
    handleSelectionChange = (e) => {
        console.log('changed!')
    }

    onEditorPreparing = (event) => {
        // console.log(event)
      
        if (
          event.parentType === "dataRow" &&
        //   event.cells.column.dataField !== 'buttons' && 
          (
            event.dataField === "id" ||
            event.dataField === "commodity" ||
            event.dataField === "quantidade" ||
            event.dataField === "preco" ||
            event.dataField === "disponibilidade"
           )
        ) {
          event.editorOptions.disabled = true;
        }
      };

      handleRowPrepared = (e) => {
          console.log(e)
        const qtdeCompra = e.newData.comprar
        const qtdeAtual = e.oldData.quantidade

          if(qtdeCompra > qtdeAtual){
            alert('valor > qtde disponível')  
            return window.location.reload()
            } else {
                this.setState({quantide: qtdeCompra})
                console.log(e.oldData)
                setOrdem ( e.oldData )
            }

      }
      refreshDataGrid() {
        this.dataGrid.instance.refresh()
      
    }

    render() {

        const {busca_dados, quantidade} = this.state.busca_dados;
    
        return (

            <>
            <div>
                <div className='std-div'>
            <h1>Perfil do Consumidor</h1>
                    <form onSubmit={this.handleSubmit}>
                <FormStd
                  width={800}
                    colCount={1}
                    formData={busca_dados}
                    onFieldDataChanged={this.handleChange}>
                        <GroupItem caption='Busca' colCount={3} >
                      
                        <SimpleItem dataField='commodity'
                            editorType='dxSelectBox' 
                            editorOptions={commoditiesOptions}/>
                        <SimpleItem dataField='estado' editorType='dxSelectBox' editorOptions={statesOptions}></SimpleItem>
                        <ButtonItem 
                        buttonOptions={buttonOptions}
                        >

                        </ButtonItem>
                       
                    </GroupItem>
                </FormStd>
                </form>
                </div>
            </div>
            <div className='std-div'>
            
                <DataGrid
                    dataSource={data}
                    ref={ref => this.dataGrid = ref}
                    defaultColumns={columns}
                    rowAlternationEnabled={true}
                    onSelectionChanged={this.handleSelectionChange}
                    onEditorPreparing = {this.onEditorPreparing}
                    // onRowPrepared = {this.handleRowPrepared}
                    onRowUpdating= {this.handleRowPrepared}
                   >
                {/* <Editing mode='popup' allowUpdating={true} useIcons={true}  /> */}
                    <Paging enabled={false} />
                    <Selection mode='multiple' />
                        <Column dataField='commodity' />
                        <Column dataField='disponibilidade' />
                        <Column dataField='quantidade' caption='Qtde disponível'  />
                        <Column dataField='preco' />
                        <Column dataField='comprar'
                            allowEditing={false} />
                <Editing mode='popup' 
                    allowUpdating={true} 
                    useIcons={true} 
                    visible={false}
                

                    >

                    <Popup title='Compra' showTitle={true} width={700} height={300} />
                    </Editing>
                    
                        {/* <Column type='buttons' dataField='executar' >
                            <Button 
                            
                                text='executar' 
                                icon='money'
                                onClick={this.handlePurchase}
                                 /> 
                        </Column> */}
                      
                            <Form>
                            
                                <Item dataField= 'commodity'/>
                                <Item dataField ='disponibilidade' />
                                <Item dataField = 'quantidade' />
                                <Item dataField = 'preco' />
                            </Form> 
                </DataGrid>
            </div>

            {/* <StdPopup
                visible={this.state.visiblePopup}
                onHiding={this.handleHide}
                showCloseButton={true}
                // closeOnOutsideClick={this.handleHide}
                showTitle={true}
                title='Compra'
                width={420}
                height={250}
                >
                    
                    <ToolbarItem text='Quantidade'
                    toolbar="bottom"
                    location="center"
                    
                    />
                  
                    <ToolbarItem widget='dxTextBox'
                    toolbar="bottom"
                    location="center"
                    key='quantidade'
                    options={this.popupQtyOptions}
                    />
                    <ToolbarItem widget='dxButton'
                    options={popupButtonOptions}
                    toolbar="bottom"
                    location="center"/>

                    <p>Preço: <span>{data[0].preco}</span></p>
                    <p>Valor: <span>{data[0].preco * quantidade}</span></p>
                    
                
            </StdPopup> */}
            </>

        

        )
    }

}

**Comment by gaprilus** [comment]
import http from './httpService'
// import {api_url} from '../config.json'

let backPath = "http://localhost:8000/produtos"
export async function setOrdem(ordem){
    return await http.post(backPath + '/cadastrar', ordem)
}

**Comment by gaprilus** [comment]
<Column dataField = 'commodity' width={100}>
              <Lookup
              
                dataSource={commodity}
                displayExpr='commodity'
                valueExpr='commodity'
              />
              <RequiredRule />
            </Column>

**Comment by gaprilus** [comment]
`   <Column dataField = 'profile' width={100}>
              <Lookup
              
                dataSource={profiles}
                displayExpr='tipo'
                valueExpr='tipo'
              />
              <RequiredRule />
            </Column>`

**Comment by gaprilus** [comment]
""
   <Column dataField = 'profile' width={100}>
              <Lookup
              
                dataSource={profiles}
                displayExpr='tipo'
                valueExpr='tipo'
              />
              <RequiredRule />
            </Column>

""

**Comment by gaprilus** [comment]
```
   <Column dataField = 'profile' width={100}>
              <Lookup
              
                dataSource={profiles}
                displayExpr='tipo'
                valueExpr='tipo'
              />
              <RequiredRule />
            </Column>

```

**Comment by gaprilus** [comment]
import React, {Component} from 'react'
import './App.css';
import 'devextreme/dist/css/dx.greenmist.css';

import {
  BrowserRouter,
  Switch,
  Route,
  Routes,
 
} from "react-router-dom";

import Home from './components/home'
import Navbar from './components/navbar'
import Cadastro from './components/cadastro'
import Produtor from './components/produtor'
import Consumidor from './components/consumidor'
import Menu from './components/menu'
import Compras from './components/compras'


class App extends Component {

  state = {
    login: {
      email: '',
      senha: '' ,

    }
  }

  getUser = () =>{
    


    console.log(this.state)
  }

  componentDidMount (){
    const login = { email: 'aluno@ufabc.com.br' , senha: 'y'}
    this.setState ({
      login:login
    })

  }

  render(){

    const user = this.getUser()
    return (
      <BrowserRouter>
    {/* <Menu/> */}
    <Navbar user= {this.state.login} />
    <Routes>
        <Route exact path='/' element={ <Home/>} />
        <Route exact path='/cadastro' element={<Cadastro/>} />
        <Route exact path='/produtor' element={<Produtor/>} />
        <Route exact path='/consumidor' element={<Consumidor/>} />
        <Route exact path='/compras' element={<Compras/>} />
      </Routes>
    </BrowserRouter>
  );
}
}

export default App;

**Comment by gaprilus** [comment]
{
  "name": "engsoft-ufabc",
  "version": "0.1.0",
  "homepage": "https://hespius.github.io/SMA-Handshake",
  "private": true,
  "dependencies": {
    "@testing-library/jest-dom": "^5.16.3",
    "@testing-library/react": "^12.1.4",
    "@testing-library/user-event": "^13.5.0",
    "axios": "^0.26.1",
    "devextreme-react": "^21.2.6",
    "gh-pages": "^3.2.3",
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "react-router-dom": "^6.2.2",
    "react-scripts": "5.0.0",
    "web-vitals": "^2.1.4",
    "yarn": "^1.22.18"
  },
  "scripts": {
    "predeploy": "npm run build",
    "deploy": "gh-pages -d build",
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test",
    "eject": "react-scripts eject"
  },
  "eslintConfig": {
    "extends": [
      "react-app",
      "react-app/jest"
    ]
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  }
}

**Comment by antunesvitor** [comment]
```
[
    {
        "id": 1,
        "data_cadastro": "2022-04-15T12:59:59.247096",
        "data_disponivel": "2022-04-08T00:00:00",
        "quantidade": 150.0,
        "preco": 9.23,
        "commodity": {
            "id": 1,
            "nome": "Arroz",
            "usuarios": []
        },
        "usuario": {
            "id": 1,
            "nome": "produtor1",
            "telefone": "11 9 1111-1111",
            "email": "produtor1@teste.com.br",
            "logradouro": "Rua Sei la",
            "numero": 123,
            "complemento": null,
            "cep": "12345-012",
            "cidade": "Santo andré",
            "estado": "SP",
            "cpf_cnpj": "123.456.789-00",
            "tipo": "Produtor",
            "senha": "senha2",
            "commodity": []
        }
    },
    {
        "id": 2,
        "data_cadastro": "2022-04-15T16:44:23.502624",
        "data_disponivel": "2022-04-08T00:00:00",
        "quantidade": 200.0,
        "preco": 10.55,
        "commodity": {
            "id": 2,
            "nome": "Feijão",
            "usuarios": []
        },
        "usuario": {
            "id": 2,
            "nome": "produtor2",
            "telefone": "11 9 1111-1111",
            "email": "produtor2@teste.com.br",
            "logradouro": "Rua Sei la",
            "numero": 123,
            "complemento": null,
            "cep": "12345-012",
            "cidade": "Santo andré",
            "estado": "SP",
            "cpf_cnpj": "123.456.789-00",
            "tipo": "Produtor",
            "senha": "senha2",
            "commodity": []
        }
    },
    {
        "id": 3,
        "data_cadastro": "2022-04-15T16:44:38.232635",
        "data_disponivel": "2022-04-08T00:00:00",
        "quantidade": 250.0,
        "preco": 20.99,
        "commodity": {
            "id": 3,
            "nome": "Trigo",
            "usuarios": []
        },
        "usuario": {
            "id": 3,
            "nome": "produtor3",
            "telefone": "11 9 1111-1111",
            "email": "produtor3@teste.com.br",
            "logradouro": "Rua Sei la",
            "numero": 123,
            "complemento": null,
            "cep": "12345-012",
            "cidade": "Santo andré",
            "estado": "SP",
            "cpf_cnpj": "123.456.789-00",
            "tipo": "Produtor",
            "senha": "senha2",
            "commodity": []
        }
    },
    {
        "id": 4,
        "data_cadastro": "2022-04-15T16:49:42.463060",
        "data_disponivel": "2022-04-08T00:00:00",
        "quantidade": 300.0,
        "preco": 15.99,
        "commodity": {
            "id": 4,
            "nome": "Tomate",
            "usuarios": []
        },
        "usuario": {
            "id": 1,
            "nome": "produtor1",
            "telefone": "11 9 1111-1111",
            "email": "produtor1@teste.com.br",
            "logradouro": "Rua Sei la",
            "numero": 123,
            "complemento": null,
            "cep": "12345-012",
            "cidade": "Santo andré",
            "estado": "SP",
            "cpf_cnpj": "123.456.789-00",
            "tipo": "Produtor",
            "senha": "senha2",
            "commodity": []
        }
    }
]
```

**Comment by antunesvitor** [comment]
```
[
    {
        "id": 1,
        "nome": "Arroz",
        "usuarios": []
    },
    {
        "id": 2,
        "nome": "Feijão",
        "usuarios": []
    },
    {
        "id": 3,
        "nome": "Trigo",
        "usuarios": []
    },
    {
        "id": 4,
        "nome": "Tomate",
        "usuarios": []
    },
    {
        "id": 5,
        "nome": "Alface",
        "usuarios": []
    },
    {
        "id": 6,
        "nome": "Cebola",
        "usuarios": []
    },
    {
        "id": 7,
        "nome": "Cenoura",
        "usuarios": []
    },
    {
        "id": 8,
        "nome": "Maçã",
        "usuarios": []
    },
    {
        "id": 9,
        "nome": "Laranja",
        "usuarios": []
    },
    {
        "id": 10,
        "nome": "Mandioca",
        "usuarios": []
    },
    {
        "id": 11,
        "nome": "Banana",
        "usuarios": []
    }
]
```

**Comment by gaprilus** [comment]
import React, {Component} from 'react';

import DataGrid, {
  Column,
  Editing,
  Popup,
  Paging,
  Lookup,
  Form,
  Item,
  Button,
  
  Selection,
} from 'devextreme-react/data-grid';
import {Form as FormStd, SimpleItem, ButtonItem, GroupItem} from 'devextreme-react/form';
import { Popup as StdPopup, Position, ToolbarItem, Item as PopupItem   } from 'devextreme-react/popup';

import {setOrdem} from '../middleware/servicesConsumidor'


const columns = ['id', 'commodity','disponibilidade', 'quantidade','preco','comprar']

const data = [{
    id: 1,
    commodity: 'soja',
    disponibilidade: '2022/05/01',
    quantidade: 10,
    preco: 90,
    comprar: 0,
    // executar: false,
    

}]

const commoditiesOptions = { items: ['Milho','Soja','Trigo'], searchEnabled: true, value: '' };
const statesOptions = { items: ['São Paulo','Bahia','Rio de Janeiro'], searchEnabled: true, value: '' };

const popupButtonOptions ={ 
    text: 'Comprar', 
    width: 100,
}

const buttonOptions={
    text: 'Buscar',
    type: 'success',
    useSubmitBehavior: true,
    width: 200
  }



export default class Consumidor extends Component {
    
  
    state = {
        busca_dados: {
        commodity: '',
        estado: '',
    },
    visiblePopup: false,
    quantidade: 0,
    }

    popupQtyOptions={
        
        onDataFieldChange: this.handleQtyChange,
        
      }
    
handleQtyChange = (e) => {
    console.log(e.value)
}

    handleSubmit = (e) => {
        e.preventDefault()
        console.log(e)
        console.log(this.state.busca_dados)
    }

    handleChange = (e) => {

        const {busca_dados} = this.state
        let newDados = busca_dados

        console.log(e.value)
       
        newDados[e.dataField] = e.value
        this.setState({busca_dados: newDados});
    
    }

    handlePurchase = (e) => {
        console.log('Compra!')

        this.setState({visiblePopup:true})
    
        
        console.log(this.state.visiblePopup)
    }

    handleHide = (e) => {
        console.log('Close!')

        this.setState({visiblePopup:false})

        console.log(this.state.visiblePopup)
    }
    handleSelectionChange = (e) => {
        console.log('changed!')
    }

    onEditorPreparing = (event) => {
        // console.log(event)
      
        if (
          event.parentType === "dataRow" &&
        //   event.cells.column.dataField !== 'buttons' && 
          (
            event.dataField === "id" ||
            event.dataField === "commodity" ||
            event.dataField === "quantidade" ||
            event.dataField === "preco" ||
            event.dataField === "disponibilidade"
           )
        ) {
          event.editorOptions.disabled = true;
        }
      };

      handleRowPrepared = (e) => {
          console.log(e)
        const qtdeCompra = e.newData.comprar
        const qtdeAtual = e.oldData.quantidade

          if(qtdeCompra > qtdeAtual){
            alert('valor > qtde disponível')  
            return window.location.reload()
            } else {
                this.setState({quantide: qtdeCompra})
                console.log(e.oldData)
                setOrdem ( e.oldData )
            }

      }
      refreshDataGrid() {
        this.dataGrid.instance.refresh()
      
    }

    render() {

        const {busca_dados, quantidade} = this.state.busca_dados;
    
        return (

            <>
            <div>
                <div className='std-div'>
            <h1>Perfil do Consumidor</h1>
                    <form onSubmit={this.handleSubmit}>
                <FormStd
                  width={800}
                    colCount={1}
                    formData={busca_dados}
                    onFieldDataChanged={this.handleChange}>
                        <GroupItem caption='Busca' colCount={3} >
                      
                        <SimpleItem dataField='commodity'
                            editorType='dxSelectBox' 
                            editorOptions={commoditiesOptions}/>
                        <SimpleItem dataField='estado' editorType='dxSelectBox' editorOptions={statesOptions}></SimpleItem>
                        <ButtonItem 
                        buttonOptions={buttonOptions}
                        >

                        </ButtonItem>
                       
                    </GroupItem>
                </FormStd>
                </form>
                </div>
            </div>
            <div className='std-div'>
            
                <DataGrid
                    dataSource={data}
                    ref={ref => this.dataGrid = ref}
                    defaultColumns={columns}
                    rowAlternationEnabled={true}
                    onSelectionChanged={this.handleSelectionChange}
                    onEditorPreparing = {this.onEditorPreparing}
                    // onRowPrepared = {this.handleRowPrepared}
                    onRowUpdating= {this.handleRowPrepared}
                   >
                {/* <Editing mode='popup' allowUpdating={true} useIcons={true}  /> */}
                    <Paging enabled={false} />
                    <Selection mode='multiple' />
                        <Column dataField='commodity' />
                        <Column dataField='disponibilidade' />
                        <Column dataField='quantidade' caption='Qtde disponível'  />
                        <Column dataField='preco' />
                        <Column dataField='comprar'
                            allowEditing={false} />
                <Editing mode='popup' 
                    allowUpdating={true} 
                    useIcons={true} 
                    visible={false}
                

                    >

                    <Popup title='Compra' showTitle={true} width={700} height={300} />
                    </Editing>
                    
                      
                            <Form>
                            
                                <SelectBox dataField= 'commodity'/>
                                <Item dataField ='disponibilidade' />
                                <Item dataField = 'quantidade' />
                                <Item dataField = 'preco' />
                            </Form> 
                </DataGrid>
            </div>

            {/* <StdPopup
                visible={this.state.visiblePopup}
                onHiding={this.handleHide}
                showCloseButton={true}
                // closeOnOutsideClick={this.handleHide}
                showTitle={true}
                title='Compra'
                width={420}
                height={250}
                >
                    
                    <ToolbarItem text='Quantidade'
                    toolbar="bottom"
                    location="center"
                    
                    />
                  
                    <ToolbarItem widget='dxTextBox'
                    toolbar="bottom"
                    location="center"
                    key='quantidade'
                    options={this.popupQtyOptions}
                    />
                    <ToolbarItem widget='dxButton'
                    options={popupButtonOptions}
                    toolbar="bottom"
                    location="center"/>

                    <p>Preço: <span>{data[0].preco}</span></p>
                    <p>Valor: <span>{data[0].preco * quantidade}</span></p>
                    
                
            </StdPopup> */}
            </>

        

        )
    }

}

**Comment by gaprilus** [comment]
import React, {Component} from 'react'
import { setLogin } from '../middleware/servicesCadastro'

export default class Navbar extends Component {

    constructor (props){
        super(props)

        this.getUser = () => {
            return this.state.login
        }
    }
state = {
    login:{
    email: '',
    senha: ' ',
    }
}



handleBlur = (e) => {
    const {login} = this.state
    let newLogin = login

    newLogin[e.target.id] = e.target.value
    this.setState({login: newLogin});
    console.log(login)
}
handleSubmit = async (e) => {
    e.preventDefault()
    const login = this.state.login
    try {
        var response = await setLogin(login)
        console.log(response)
        this.loginRedirect(response)
    }
    catch (error){
        alert(error.response.data.detail)
    }
}

loginRedirect = (response) => {
    if(response.status === 200){
        if(response.data.tipo === "Produtor"){
            window.location.href = "/produtor"
        }
        else {
            window.location.href = "/consumidor"
        }
    }
    else {
        alert("Erro ao efetuar o login")
    }
}


render() {

    const { user } = this.props;    
    
    return (

        <>
        <header>
        <div>
            <h1><a href='/'>SMA Handshake</a></h1>
        </div>
        {
            !user && (
            
            <div className='login-div'>
            <form onSubmit={this.handleSubmit}>
            <label>Usuário</label>
            <input id='email' onBlur={this.handleBlur}></input>
            <label>Senha</label>
            <input id='senha' type="password" onBlur={this.handleBlur}></input>
            <button >Entrar</button>
            </form>
            <a href='cadastro'>Cadastrar-se</a>
        </div>
        )
        }
        {
            user && (
                <div >
                    <p>{user.email}</p>
                </div>
            )
        }
        </header>
        </>

    )
}

}

**Comment by antunesvitor** [comment]
Endpoint proposto para criação  e buscar de ordens:
POST /produtos/ordem
BODY (exemplo)
```
{
    "quantidade": 10.0,
    "data_requisitada": "2022-04-22T00:00:00",
    "comprador": 1,
    "oferta": 7
}
```

RESPONSE (exemplo)
```
{
    "id": 3,
    "quantidade": 10.0,
    "data_requisitada": "2022-04-22T00:00:00",
    "oferta": {
        "id": 7
    },
    "comprador": {
        "id": 1,
        "oferta": null
    }
}
```

GET /produtos/ordem?
params:
consumidorId: número inteiro
produtorId: número inteiro
commodityId: número inteiro [não implementado ainda]

RESPONSE:
```
[
    {
        "id": 3,
        "quantidade": 10.0,
        "data_requisitada": "2022-04-22T00:00:00",
        "comprador": {
            "id": 1,
            "nome": "produtor1",
            "telefone": "11 9 1111-1111",
            "email": "produtor1@teste.com.br",
            "logradouro": "Rua Sei la",
            "numero": 123,
            "complemento": null,
            "cep": "12345-012",
            "cidade": "Santo andré",
            "estado": "SP",
            "cpf_cnpj": "123.456.789-00",
            "tipo": "Produtor",
            "senha": "senha",
            "commodity": []
        },
        "oferta": {
            "id": 7,
            "data_cadastro": "2022-04-21T22:18:37.153716",
            "data_disponivel": "2022-04-08T00:00:00",
            "quantidade": 80.0,
            "saldo": 70.0,
            "preco": 39.99,
            "usuario": {
                "id": 2,
                "nome": "produtor2",
                "telefone": "11 9 1111-1111",
                "email": "produtor2@teste.com.br",
                "logradouro": "Rua Sei la",
                "numero": 123,
                "complemento": null,
                "cep": "12345-012",
                "cidade": "Santo andré",
                "estado": "SP",
                "cpf_cnpj": "123.456.789-00",
                "tipo": "Produtor",
                "senha": "senha",
                "commodity": []
            },
            "commodity": {
                "id": 7,
                "oferta": null
            }
        }
    }
]
```

**Comment by gaprilus** [comment]
import React, {Component} from 'react';

import DataGrid, {
  Column,
  Editing,
  Popup,
  Paging,
  Lookup,
  Form,
} from 'devextreme-react/data-grid';
import 'devextreme-react/text-area';
import { Item } from 'devextreme-react/form';
import { setProduto } from '../middleware/servicesProduto';

const data = [
    {
        "id": 1,
        "quantidade": 99.99,
        "data_requisitada": "2022-04-18T00:00:00",
        "usuario": 1,
        "commodity": "Milho"
    }
  ]

export default class Produtor extends Component {
    

    constructor(props){
        super(props)
        const commodities= [{commodity: 'milho'}, {commodity: 'soja'}]  
    }
    state = {
        cadastro_venda: {
        commodity: '',
        data_disponivel: '',
        quantidade: 0,
        preco: 0,
    },
    commodities: [{commodity: 'milho'}, {commodity: 'soja'}]
    }
    handleSubmit = (e) => {
        e.preventDefault()
        console.log('submit')
    }

    onInsertHandler = async (ev) => {
        console.log(ev.data)
        const data = {commodity: ev.data.commodity, preco: ev.data.preco, quantidade: ev.data.quantidade, data_disponivel: ev.data.data_disponivel}
        console.log(data)
        var response = await setProduto(data)
        console.log(response)
    }


    render() {

        const {commodities} = this.state
        return (

            <>
            <div className='std-div'>
            <h1>Perfil do Produtor</h1>
                <DataGrid
                    dataSource={data}
                    onRowInserting={this.onInsertHandler}>
                    <Paging enabled={false} />
                            

                            <Popup title='Registrar' showTitle={true} width={700} height={500} />
                            <Column dataField='id' />
                            <Column dataField='quantidade' />
                            <Column dataField='usuario' />
                            <Column dataField="data_requisitada" dataType="date"></Column>
                            <Column dataField='commodity'>
                            <Lookup
                            allowClearing={true}
                            dataSource={commodities}
                            displayExpr='commodity'
                            valueExpr='commodity'
                            /></Column>
                    <Editing
                        mode='popup'
                        allowAdding={true}
                        allowUpdating={true}
                        allowDeleting={true}
                        useIcons={true}>
                            <Form>
                                <Item dataField = 'commodity' />
                                <Item dataField = 'data_requisitada' dataType='date'/>
                                <Item dataField = 'quantidade' />
                                <Item dataField = 'preco' />

                            </Form>
                        </Editing>
                </DataGrid>
            </div>
            </>

        

        )
    }

}

**Comment by gaprilus** [comment]
dataSource = {[{commodity: 'milho'}, {commodity:'trigo'}]}
